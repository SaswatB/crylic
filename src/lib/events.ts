import { createEventDefinition, EventBus } from "ts-bus";

export const bus = new EventBus();

export const editorResize = createEventDefinition()("editor.resize");
