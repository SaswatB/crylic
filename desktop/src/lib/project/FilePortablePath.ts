import { normalizePath } from "synergy/src/lib/normalizePath";
import { PortablePath } from "synergy/src/lib/project/PortablePath";

const pathModule = __non_webpack_require__("path") as typeof import("path");

export class FilePortablePath extends PortablePath {
  public constructor(protected nativePath: string) {
    super();
  }

  public override isEqual(path: PortablePath): boolean {
    if (!(path instanceof FilePortablePath))
      throw new Error("Unsupported PortablePath type");
    return this.nativePath === path.nativePath;
  }

  public override join(path: string): PortablePath {
    return new FilePortablePath(
      pathModule.join(this.nativePath, normalizePath(path, pathModule.sep))
    );
  }
  public override relative(to: PortablePath): PortablePath {
    if (!(to instanceof FilePortablePath))
      throw new Error("Unsupported PortablePath type");
    return new FilePortablePath(
      pathModule.relative(this.nativePath, to.nativePath)
    );
  }

  public override getDirname(): PortablePath {
    return new FilePortablePath(pathModule.dirname(this.nativePath));
  }
  public override getBasename(): string {
    return pathModule.basename(this.nativePath);
  }

  public override getNativePath(): string {
    return this.nativePath;
  }
  public override getNormalizedPath(): string {
    return normalizePath(this.nativePath, "/");
  }
}
