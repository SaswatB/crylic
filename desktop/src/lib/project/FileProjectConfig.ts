import { fold } from "fp-ts/lib/Either";
import { pipe } from "fp-ts/lib/pipeable";

import { CONFIG_FILE_NAME } from "synergy/src/constants";
import { PackageManager } from "synergy/src/lib/pkgManager/PackageManager";
import { PortablePath } from "synergy/src/lib/project/PortablePath";
import {
  ProjectConfig,
  ProjectConfigFile,
} from "synergy/src/lib/project/ProjectConfig";

import { requireUncached } from "../../utils/utils";
import { InbuiltPackageManager } from "../pkgManager/InbuiltPackageManager";

const fs = __non_webpack_require__("fs") as typeof import("fs");

export class FileProjectConfig extends ProjectConfig {
  public static createProjectConfigFromDirectory(projectPath: PortablePath) {
    let configFile;
    let packageJson;

    // process the config file
    const configFilePath = projectPath.join(CONFIG_FILE_NAME);
    if (fs.existsSync(configFilePath.getNativePath())) {
      // todo use a more secure require/allow async
      configFile = pipe(
        configFilePath.getNativePath(),
        // require the config file
        requireUncached,
        // parse the config file
        ProjectConfigFile.decode,
        fold(
          // log any errors
          (e) => {
            console.log("ProjectConfigFile decode error", e);
            return undefined;
          },
          (config) => config
        )
      );
    }

    // process package.json
    const packageJsonPath = projectPath.join("package.json");
    if (fs.existsSync(packageJsonPath.getNativePath())) {
      packageJson = JSON.parse(
        fs.readFileSync(packageJsonPath.getNativePath(), { encoding: "utf-8" })
      );
    }

    return new FileProjectConfig(projectPath, configFile || {}, packageJson);
  }

  private packageManager: PackageManager | undefined;
  public getPackageManager() {
    if (!this.packageManager) {
      const packageManagerType =
        this.configFile?.packageManager?.type ?? "inbuilt";

      if (packageManagerType.startsWith("inbuilt")) {
        let yarnMode = false;
        if (packageManagerType === "inbuilt-yarn") {
          yarnMode = true;
        } else if (packageManagerType !== "inbuilt-npm") {
          if (
            fs.existsSync(this.projectPath.join("yarn.lock").getNativePath())
          ) {
            yarnMode = true;
          } else {
            try {
              if (this.packageJson.packageManager?.includes("yarn"))
                yarnMode = true;
            } catch (e) {
              // ignore
            }
          }
        }
        this.packageManager = new InbuiltPackageManager(
          this.projectPath,
          yarnMode
        );
      } else if (packageManagerType === "yarn") {
        // yarn
        throw new Error("Not implemented");
      } else {
        // npm
        throw new Error("Not implemented");
      }
    }

    return this.packageManager;
  }
}
