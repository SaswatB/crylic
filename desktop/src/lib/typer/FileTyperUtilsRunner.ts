import * as Comlink from "comlink";
import { Subscription } from "rxjs";
import { distinctUntilChanged, map } from "rxjs/operators";

import { LTBehaviorSubject } from "synergy/src/lib/lightObservable/LTBehaviorSubject";
import { CodeEntry } from "synergy/src/lib/project/CodeEntry";
import { PortablePath } from "synergy/src/lib/project/PortablePath";
import { eagerMapArrayAny } from "synergy/src/lib/rxjs/eagerMap";

import type { FileTyperUtilsWorker } from "./FileTyperUtilsWorker";

export class FileTyperUtilsRunner {
  private worker = new Worker(
    new URL("./FileTyperUtilsWorker.ts", import.meta.url)
  );
  private proxy = Comlink.wrap<FileTyperUtilsWorker>(this.worker);
  private codeEntrySubscription: Subscription;

  constructor(
    protected projectPath: PortablePath,
    protected readonly codeEntries$: LTBehaviorSubject<CodeEntry[]>
  ) {
    // init the ts service
    void this.proxy.init(
      projectPath.getNativePath(),
      codeEntries$.getValue().map((c) => c.getRemoteCodeEntry())
    );

    // keep track of code updates and send them to the worker
    this.codeEntrySubscription = this.codeEntries$
      .toRXJS()
      .pipe(
        eagerMapArrayAny((codeEntry) =>
          codeEntry.code$.toRXJS().pipe(
            distinctUntilChanged(),
            map(() => codeEntry)
          )
        )
      )
      .subscribe((codeEntry) => {
        // todo debounce?
        void this.proxy.updateCodeEntries([codeEntry.getRemoteCodeEntry()]);
      });
  }

  public getProxy() {
    return this.proxy;
  }

  public dispose() {
    this.codeEntrySubscription.unsubscribe();
    this.proxy[Comlink.releaseProxy]();
    this.worker.terminate();
  }
}
