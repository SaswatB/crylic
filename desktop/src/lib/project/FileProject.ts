import { MakeDirectoryOptions } from "fs";
import minimatch from "minimatch";
import { Readable } from "stream";
import yauzl from "yauzl";

import { track } from "synergy/src/hooks/useTracking";
import { bus, fileSyncConflict, fileSyncSuccess } from "synergy/src/lib/events";
import {
  IMAGE_EXTENSION_REGEX,
  SCRIPT_EXTENSION_REGEX,
  STYLE_EXTENSION_REGEX,
} from "synergy/src/lib/ext-regex";
import { normalizePath } from "synergy/src/lib/normalizePath";
import {
  CodeEntry,
  INITIAL_CODE_REVISION_ID,
} from "synergy/src/lib/project/CodeEntry";
import { PortablePath } from "synergy/src/lib/project/PortablePath";
import { Project } from "synergy/src/lib/project/Project";
import { ProjectConfig } from "synergy/src/lib/project/ProjectConfig";
import { sleep } from "synergy/src/lib/utils";
import { PluginService } from "synergy/src/services/PluginService";

import { streamToString } from "../../utils/utils";
import { FileTyperUtilsRunner } from "../typer/FileTyperUtilsRunner";
import { FileProjectConfig } from "./FileProjectConfig";

const fs = __non_webpack_require__("fs") as typeof import("fs");
const fsP = fs.promises;
const chokidar = __non_webpack_require__(
  "chokidar"
) as typeof import("chokidar");

export enum FileProjectTemplate {
  Blank = "blank",
}

const TemplateBuffers: Record<FileProjectTemplate, () => Promise<Buffer>> = {
  [FileProjectTemplate.Blank]: () =>
    import(
      "!!../../../loaders/binaryLoader!../../assets/project-blank-template.zip"
    ).then((b) => b.default),
};

export class FileProject extends Project {
  public static async createNewProjectInDirectory(
    folderPath: PortablePath,
    template: FileProjectTemplate,
    pluginService: PluginService
  ) {
    console.log(folderPath);
    if (
      !(await fsP
        .access(folderPath.getNativePath())
        .then(() => true)
        .catch(() => false))
    )
      await fsP.mkdir(folderPath.getNativePath(), { recursive: true });

    const buffer = await TemplateBuffers[template]();

    let canceled = false;
    await new Promise<void>((resolve, reject) => {
      yauzl.fromBuffer(buffer, {}, (err, zipFile) => {
        if (err || !zipFile) throw new Error("Failed to read project template");

        zipFile.on("error", (err) => {
          canceled = true;
          reject(err);
        });

        let readCount = 0;
        zipFile.on("entry", async (entry) => {
          if (canceled) return;
          console.log("zip entry", entry);

          try {
            const dest = folderPath.join(entry.fileName);

            // convert external file attr int into a fs stat mode int
            const mode = (entry.externalFileAttributes >> 16) & 0xffff;
            // check if it's a symlink or dir (using stat mode constants)
            const IFMT = 61440;
            const IFDIR = 16384;
            const IFLNK = 40960;
            const symlink = (mode & IFMT) === IFLNK;
            let isDir = (mode & IFMT) === IFDIR;

            // Failsafe, borrowed from jsZip
            if (!isDir && entry.fileName.endsWith("/")) {
              isDir = true;
            }

            // check for window's weird way of specifying a directory
            // https://github.com/maxogden/extract-zip/issues/13#issuecomment-154494566
            const madeBy = entry.versionMadeBy >> 8;
            if (!isDir)
              isDir = madeBy === 0 && entry.externalFileAttributes === 16;

            // reverse umask first (~)
            const umask = ~process.umask();
            // & with processes umask to override invalid perms
            const procMode = (mode || (isDir ? 0o755 : 0o644)) & umask;

            // always ensure folders are created
            const destDir = isDir ? dest : dest.getDirname();

            const mkdirOptions: MakeDirectoryOptions = { recursive: true };
            if (isDir) {
              mkdirOptions.mode = procMode;
            }
            await fsP.mkdir(destDir.getNativePath(), mkdirOptions);
            if (isDir) return;

            const readStream = await new Promise<Readable>(
              (subResolve, subReject) =>
                zipFile.openReadStream(entry, (subErr, stream) => {
                  if (subErr || !stream)
                    subReject(new Error("Failed to read zip entry"));
                  else subResolve(stream);
                })
            );

            if (symlink) {
              await fsP.symlink(
                await streamToString(readStream),
                dest.getNativePath()
              );
            } else {
              readStream.pipe(
                fs.createWriteStream(dest.getNativePath(), { mode: procMode })
              );
            }

            if (++readCount === zipFile.entryCount) {
              resolve();
            }
          } catch (err) {
            canceled = true;
            reject(err);
          }
        });
      });
    });

    track("project.create", { template });
    // sleep to flush changes and avoid blank loading bug
    await sleep(1000);

    return FileProject.createProjectFromDirectory(folderPath, pluginService);
  }

  public static async createProjectFromDirectory(
    folderPath: PortablePath,
    pluginService: PluginService
  ) {
    // build metadata
    const refreshConfig = () => {
      const newConfig: ProjectConfig =
        FileProjectConfig.createProjectConfigFromDirectory(folderPath);
      pluginService.activatePlugins(newConfig);
      return pluginService.reduceActive(
        (acc, p) => p.overrideProjectConfig(acc, { fs }),
        newConfig
      );
    };
    const config: ProjectConfig = refreshConfig();
    const project = new FileProject(folderPath, config, refreshConfig);

    // process all the source files
    const fileCodeEntries: CodeEntry[] = [];
    // recursive function for creating code entries from a folder
    const read = async (subFolderPath: PortablePath) =>
      Promise.all(
        (await fsP.readdir(subFolderPath.getNativePath())).map(async (file) => {
          const filePath = subFolderPath.join(file);
          // skip ignored paths
          if (
            config
              .getIgnoredPaths()
              .find((g) =>
                minimatch(
                  filePath
                    .getNormalizedPath()
                    .replace(folderPath.getNormalizedPath(), ""),
                  g
                )
              )
          ) {
            console.log("ignoring file due to config", filePath);
            return;
          }
          if ((await fsP.stat(filePath.getNativePath())).isDirectory()) {
            // recurse through the directory's children
            await read(filePath);
          } else if (
            file.match(SCRIPT_EXTENSION_REGEX) ||
            file.match(STYLE_EXTENSION_REGEX)
          ) {
            // add scripts/styles as code entries
            const code = await fsP.readFile(filePath.getNativePath(), {
              encoding: "utf-8",
            });
            fileCodeEntries.push(new CodeEntry(project, filePath, code));
          } else if (file.match(IMAGE_EXTENSION_REGEX)) {
            // add images as code entries without text code
            fileCodeEntries.push(new CodeEntry(project, filePath, undefined));
          } else {
            console.log("ignoring file due to extension", filePath);
          }
        })
      );
    // read all files within the folder
    await read(folderPath);

    // set the window name
    document.title = `${config.name} - Crylic`;

    project.addCodeEntries(fileCodeEntries);
    project.getTyperUtils(); // load typer utils after all code entries are loaded
    return project;
  }

  private fileWatcher: import("chokidar").FSWatcher;
  private fileChangeQueue = new Set<string>();
  private fileChangeQueueTimer: number | undefined;

  protected constructor(
    path: PortablePath,
    config: ProjectConfig,
    private refreshConfigImpl: () => ProjectConfig
  ) {
    super(path, config);

    // watch for changes on files in the source folder
    this.fileWatcher = chokidar
      .watch(path.getNativePath(), {
        ignored: (p) =>
          !!config
            .getIgnoredPaths()
            .find((g) =>
              minimatch(
                normalizePath(p.replace(path.getNativePath(), ""), "/"),
                g
              )
            ),
      })
      .on("change", (changePath) => {
        this.fileChangeQueue.add(changePath);
        clearTimeout(this.fileChangeQueueTimer);
        this.fileChangeQueueTimer = setTimeout(
          () => this.processFileChangeQueue(),
          1000
        ) as unknown as number;
      });
  }

  public override onClose() {
    super.onClose();
    void this.fileWatcher.close();
    clearTimeout(this.fileChangeQueueTimer);
    this.fileTyperUtilsRunner?.dispose();
  }

  private savedCodeRevisions: Record<
    string /* codeEntry.id */,
    { id: number; code?: string }
  > = {};
  private isCodeEntrySaved(codeEntry: CodeEntry) {
    return (
      codeEntry.codeRevisionId === INITIAL_CODE_REVISION_ID ||
      codeEntry.codeRevisionId === this.savedCodeRevisions[codeEntry.id]?.id
    );
  }

  public saveFiles() {
    this.codeEntries$
      .getValue()
      .filter(
        (e) => e.code$.getValue() !== undefined && !this.isCodeEntrySaved(e)
      )
      .forEach((e) => this.saveFile(e));
    this.projectSaved$.next();
    track("project.saved");
  }
  public saveFile({ id, filePath, code$, codeRevisionId }: CodeEntry) {
    fs.mkdirSync(filePath.getDirname().getNativePath(), { recursive: true });
    const code = code$.getValue();
    fs.writeFileSync(filePath.getNativePath(), code || "");
    this.savedCodeRevisions[id] = { id: codeRevisionId, code };
  }

  private async processFileChangeQueue() {
    this.fileChangeQueueTimer = undefined;
    const filePaths = Array.from(this.fileChangeQueue);
    this.fileChangeQueue.clear();
    const codeEntries = this.codeEntries$.getValue();

    const fileSyncSuccessPaths: string[] = [];
    const fileSyncConflictPaths: string[] = [];
    for (const filePath of filePaths) {
      const codeEntry = codeEntries.find(
        (entry) => entry.filePath.getNativePath() === filePath
      );
      if (!codeEntry) continue; // skip files that aren't tracked
      const code = codeEntry.code$.getValue();
      if (code === undefined) continue; // skip files without code
      const newCode = fs.readFileSync(filePath, { encoding: "utf-8" });
      if (newCode === code) continue; // skip unchanged files

      if (!this.isCodeEntrySaved(codeEntry)) {
        if (newCode === this.savedCodeRevisions[codeEntry.id]?.code) continue; // skip files that match the last save

        // in this case we have a file that was changed on disk and within Crylic, but Crylic did not save it
        // which is a conflict
        fileSyncConflictPaths.push(filePath);
        console.error(
          "file changed in memory and on disk (conflict)",
          filePath
        );
        continue;
      }

      // last case, update the in-memory file
      codeEntry.updateCode(newCode);
      this.savedCodeRevisions[codeEntry.id] = {
        id: codeEntry.codeRevisionId,
        code: newCode,
      };
      fileSyncSuccessPaths.push(filePath);
      console.info("file changed on disk and updated in-memory", filePath);
    }
    if (fileSyncSuccessPaths.length) {
      bus.publish(fileSyncSuccess({ paths: fileSyncSuccessPaths }));
      track("project.fileSyncSuccess", { count: fileSyncSuccessPaths.length });
    }
    if (fileSyncConflictPaths.length) {
      bus.publish(fileSyncConflict({ paths: fileSyncConflictPaths }));
      track("project.fileSyncConflict", {
        count: fileSyncConflictPaths.length,
      });
    }
  }

  public refreshConfig() {
    this.config$.next(this.refreshConfigImpl());
  }

  private fileTyperUtilsRunner: FileTyperUtilsRunner | undefined;
  public getTyperUtils() {
    if (!this.fileTyperUtilsRunner)
      this.fileTyperUtilsRunner = new FileTyperUtilsRunner(
        this.path,
        this.codeEntries$
      );
    return this.fileTyperUtilsRunner.getProxy();
  }

  public addAsset(source: PortablePath) {
    const getAssetPath = (name: string) =>
      this.path.join(`${this.config.getDefaultNewAssetFolder()}/${name}`);
    const fileName = source.getBasename();
    const assetPath = { path: getAssetPath(fileName) };
    let counter = 1;
    while (
      this.codeEntries$
        .getValue()
        .find((entry) => entry.filePath.isEqual(assetPath.path))
    ) {
      assetPath.path = getAssetPath(fileName.replace(/\./, `-${counter++}.`));
    }
    fs.writeFileSync(
      assetPath.path.getNativePath(),
      fs.readFileSync(source.getNativePath(), { encoding: null }),
      { encoding: null }
    );
    this.addCodeEntries([new CodeEntry(this, assetPath.path, undefined)]);
  }
}
