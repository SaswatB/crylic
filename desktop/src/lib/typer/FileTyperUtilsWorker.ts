import * as Comlink from "comlink";

import { RemoteCodeEntry } from "synergy/src/lib/project/CodeEntry";
import { TyperUtils } from "synergy/src/lib/typer/TyperUtils";

import { FilePortablePath } from "../project/FilePortablePath";

const ts = __non_webpack_require__("typescript") as typeof import("typescript");

let typerUtils: TyperUtils | undefined;
const codeEntries: RemoteCodeEntry[] = [];
const bootstrap = {
  init(projectPath: string, newCodeEntries: RemoteCodeEntry[]) {
    codeEntries.push(...newCodeEntries);
    typerUtils = new TyperUtils(
      new FilePortablePath(projectPath),
      codeEntries,
      ts.sys
    );
  },
  updateCodeEntries(newCodeEntries: RemoteCodeEntry[]) {
    // lm_9dfd4feb9b this doesn't support delete as is
    for (const entry of newCodeEntries) {
      const index = codeEntries.findIndex((e) => e.id === entry.id);
      if (index !== -1) codeEntries[index] = entry;
      codeEntries.push(entry);
    }
  },
};

export type FileTyperUtilsWorker = typeof bootstrap & TyperUtils;

const worker = new Proxy(bootstrap, {
  get(target, propKey) {
    if (propKey in target) return target[propKey as keyof typeof target];

    const value = typerUtils?.[propKey as keyof TyperUtils];
    if (typeof value === "function") return value.bind(typerUtils);
    return value;
  },
});

Comlink.expose(worker);
