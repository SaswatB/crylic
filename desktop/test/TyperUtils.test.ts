import { CodeEntry } from "synergy/src/lib/project/CodeEntry";
import { TyperUtils } from "synergy/src/lib/typer/TyperUtils";

import { FilePortablePath } from "../src/lib/project/FilePortablePath";
import { TestProject } from "./lib/test-utils";

function matchSnapshotProps(
  exportTarget: { name: string | undefined; isDefault: boolean },
  ...code: string[]
) {
  const props = new TyperUtils(
    "/",
    code.map((c, index) =>
      new CodeEntry(
        new TestProject(),
        new FilePortablePath(`/file${index || ""}.tsx`),
        c
      ).getRemoteCodeEntry()
    )
  ).getExportedComponentProps("/file.tsx", exportTarget);
  expect(props).toMatchSnapshot();
}

describe("TyperUtils tests", () => {
  describe("test type extractor", () => {
    test(" array prop", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export function MyComponent({ test }: { test: string[] }) {return <div />;}`
      );
    });
    test("extracts union prop", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export function MyComponent(props: { test?: string | number }) {return <div />;}`
      );
    });
    test("extracts string literal prop", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export function MyComponent(props: { test?: 'a' | 'b' }) {return <div />;}`
      );
    });
    test("extracts tuple prop", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export function MyComponent(props: { test?: [1, 2] }) {return <div />;}`
      );
    });
    test("extracts function prop", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export function MyComponent(props: { test?: () => void }) {return <div />;}`
      );
    });
    test("extracts enum prop", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `enum TestEnum { A, B }; export function MyComponent(props: { test?: TestEnum }) {return <div />;}`
      );
    });
  });

  describe("test type declaration follower", () => {
    test("get props without a type annotation", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: true },
        `export default function MyComponent({ test }) {return <div />;}`
      );
    });
    test("get props with an inline type definition", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: true },
        `export default function MyComponent(props: { test: string }) {return <div />;}`
      );
    });
    test("get props with a type definition in a different file", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: true },
        `import { Props } from './file1'; export default function MyComponent(props: Props) {return <div />;}`,
        `export type Props = { test: string };`
      );
    });
  });

  describe("test export declaration follower", () => {
    test("get props from default function", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: true },
        `export default function MyComponent({ test }: { test: string }) {return <div />;}`
      );
    });
    test("get props from default lambda function", () => {
      matchSnapshotProps(
        { name: undefined, isDefault: true },
        `export default ({ test }: { test: string }) => {return <div />;}`
      );
    });
    test("get props from named lambda function", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export const MyComponent = ({ test }: { test: string }) => {return <div />;};`
      );
    });
    test("get props from named function", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `export function MyComponent({ test }: { test: string }) {return <div />;}`
      );
    });
    test("get props from separate default function", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: true },
        `function MyComponent({ test }: { test: string }) {return <div />;}; export default MyComponent;`
      );
    });
    test("get props from separate default lambda", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: true },
        `const MyComponent = ({ test }: { test: string }) => {return <div />;}; export default MyComponent;`
      );
    });
    test("get props from separate named function", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `function MyComponent({ test }: { test: string }) {return <div />;}; export { MyComponent };`
      );
    });
    test("get props from separate named lambda", () => {
      matchSnapshotProps(
        { name: "MyComponent", isDefault: false },
        `const MyComponent = ({ test }: { test: string }) => {return <div />;}; export { MyComponent };`
      );
    });
  });
});
