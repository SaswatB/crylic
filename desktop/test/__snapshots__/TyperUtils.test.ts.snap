// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TyperUtils tests test export declaration follower get props from default function 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from default lambda function 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from named function 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from named lambda function 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from separate default function 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from separate default lambda 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from separate named function 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test export declaration follower get props from separate named lambda 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type declaration follower get props with a type definition in a different file 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type declaration follower get props with an inline type definition 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "string",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type declaration follower get props without a type annotation 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "unknown",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type extractor  array prop 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": false,
      "type": Object {
        "kind": "array",
        "memberType": Object {
          "kind": "string",
        },
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type extractor extracts enum prop 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": true,
      "type": Object {
        "kind": "union",
        "memberTypes": Array [
          Object {
            "kind": "literalNumber",
            "value": 0,
          },
          Object {
            "kind": "literalNumber",
            "value": 1,
          },
        ],
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type extractor extracts function prop 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": true,
      "type": Object {
        "kind": "function",
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type extractor extracts string literal prop 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": true,
      "type": Object {
        "kind": "union",
        "memberTypes": Array [
          Object {
            "kind": "literalString",
            "value": "a",
          },
          Object {
            "kind": "literalString",
            "value": "b",
          },
        ],
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type extractor extracts tuple prop 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": true,
      "type": Object {
        "kind": "tuple",
        "memberTypes": Array [
          Object {
            "kind": "literalNumber",
            "value": 1,
          },
          Object {
            "kind": "literalNumber",
            "value": 2,
          },
        ],
      },
    },
  ],
}
`;

exports[`TyperUtils tests test type extractor extracts union prop 1`] = `
Object {
  "kind": "object",
  "props": Array [
    Object {
      "name": "test",
      "optional": true,
      "type": Object {
        "kind": "union",
        "memberTypes": Array [
          Object {
            "kind": "string",
          },
          Object {
            "kind": "number",
          },
        ],
      },
    },
  ],
}
`;
